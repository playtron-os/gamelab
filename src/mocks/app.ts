import {
  AppDownloadStage,
  AppInformation,
  AppProvider,
  AppType,
  DownloadProgress,
  InstallConfig,
  InstalledApp,
  OwnedApp,
  PlaytronApp,
  PlaytronAppType,
  PlaytronProvider
} from "@/types";

const STEAM_OWNED_APP_MOCKS: OwnedApp[] = [
  {
    provider: AppProvider.Steam,
    name: "Apex Legends",
    id: "id1",
    provider_id: "Steam_93991",
    app_type: AppType.APP
  },
  {
    provider: AppProvider.Steam,
    name: "Counter-strike: Global Offensive",
    id: "id2",
    provider_id: "Steam_93992",
    app_type: AppType.APP
  }
];

const STEAM_DOWNLOAD_STATUS_DONE_MOCK: DownloadProgress = {
  stage: AppDownloadStage.DONE,
  progress: 100,
  error: "",
  error_code: 0,
  requires_auth: false,
  requires_2fa: false
};

const STEAM_INSTALL_CONFIG_MOCK: InstallConfig = {
  version: "1.0.0",
  install_folder: "Program Files",
  install_root: "/test/path1/root",
  install_disk: "/test/path1",
  disk_size: 100,
  download_size: 100,
  executable: "executable",
  symlink_path: "symlinkPath"
};

const STEAM_INSTALLED_APP_MOCK: InstalledApp = {
  user_id: "user_id",
  owned_app: STEAM_OWNED_APP_MOCKS[0],
  download_status: STEAM_DOWNLOAD_STATUS_DONE_MOCK,
  install_config: STEAM_INSTALL_CONFIG_MOCK,
  post_install: [],
  created_at: 0,
  updated_at: 0,
  launched_at: 0,
  interacted_at: 0
};

const STEAM_PLAYTRON_PROVIDER_MOCK: PlaytronProvider = {
  provider: AppProvider.Steam,
  providerAppId: "Steam_93991",
  lastImportedTimestamp: new Date(),
  relatedAppIds: ["id"]
};

const EPIC_PLAYTRON_PROVIDER_MOCK: PlaytronProvider = {
  provider: AppProvider.EpicGames,
  providerAppId: "EpicGames_93991",
  lastImportedTimestamp: new Date(),
  relatedAppIds: ["id"]
};

const PLAYTRON_APP_MOCKS: PlaytronApp[] = [
  {
    id: "id",
    name: "Apex Legends",
    providers: [STEAM_PLAYTRON_PROVIDER_MOCK, EPIC_PLAYTRON_PROVIDER_MOCK],
    slug: "slug",
    summary: "summary",
    description: "description",
    tags: [],
    images: [],
    publishers: [],
    developers: [],
    gameSeries: undefined,
    appType: PlaytronAppType.Game,
    releaseDate: undefined,
    comingSoon: false,
    ageRating: undefined,
    credits: undefined,
    onlyPlaytron: false,
    popularity: [],
    ratings: [],
    sales: [],
    createdAt: new Date(),
    updatedAt: new Date(),
    cmsId: 0
  },
  {
    id: "test",
    name: "Counter-strike: Global Offensive",
    providers: [STEAM_PLAYTRON_PROVIDER_MOCK, EPIC_PLAYTRON_PROVIDER_MOCK],
    slug: "slug",
    summary: "summary",
    description: "description",
    tags: [],
    images: [],
    publishers: [],
    developers: [],
    gameSeries: undefined,
    appType: PlaytronAppType.Game,
    releaseDate: undefined,
    comingSoon: false,
    ageRating: undefined,
    credits: undefined,
    onlyPlaytron: false,
    popularity: [],
    ratings: [],
    sales: [],
    createdAt: new Date(),
    updatedAt: new Date(),
    cmsId: 0
  }
];

const EPIC_OWNED_APP_MOCKS: OwnedApp[] = [
  {
    provider: AppProvider.EpicGames,
    name: "Fortnite",
    id: "xe23ft",
    provider_id: "EpicGames_Fortnite",
    app_type: AppType.APP
  }
];

export const STEAM_APP_INFORMATION_MOCKS: AppInformation[] = [
  {
    app: PLAYTRON_APP_MOCKS[0],
    installed_app: STEAM_INSTALLED_APP_MOCK,
    is_downloading: false,
    is_installing: false,
    is_launched: false,
    is_owned: true,
    is_paused: false,
    is_running: false,
    owned_apps: [STEAM_OWNED_APP_MOCKS[0], EPIC_OWNED_APP_MOCKS[0]]
  },
  {
    app: PLAYTRON_APP_MOCKS[1],
    installed_app: STEAM_INSTALLED_APP_MOCK,
    is_downloading: false,
    is_installing: false,
    is_launched: false,
    is_owned: true,
    is_paused: false,
    is_running: false,
    owned_apps: [STEAM_OWNED_APP_MOCKS[0], EPIC_OWNED_APP_MOCKS[0]]
  }
];
