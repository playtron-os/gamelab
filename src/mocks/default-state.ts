export const DEFAULT_STATE_MOCK = {
  auth: {
    userId: "userId"
  }
};
