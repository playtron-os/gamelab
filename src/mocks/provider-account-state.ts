import { AccountsState, ProviderAccountState } from "@/redux/modules";
import { AppProvider } from "@/types";

export const CONNECTED_PROVIDER_ACCOUNT_STATE: ProviderAccountState = {
  isConnected: true,
  isLoading: false,
  error: null,
  requiresTwoFactorCode: false
};

export const DISCONNECTED_PROVIDER_ACCOUNT_STATE: ProviderAccountState = {
  isConnected: false,
  isLoading: false,
  error: null,
  requiresTwoFactorCode: false
};

export const LOADING_PROVIDER_ACCOUNT_STATE: ProviderAccountState = {
  isConnected: false,
  isLoading: true,
  error: null,
  requiresTwoFactorCode: false
};

export const ERROR_PROVIDER_ACCOUNT_STATE: ProviderAccountState = {
  isConnected: false,
  isLoading: false,
  error: "Error",
  requiresTwoFactorCode: false
};

export const TWO_FACTOR_PROVIDER_ACCOUNT_STATE: ProviderAccountState = {
  isConnected: false,
  isLoading: false,
  error: null,
  requiresTwoFactorCode: true
};

export const PROVIDERS_ACCOUNTS_DISCONNECTED_STATE_MOCK: AccountsState = {
  providers: {
    [AppProvider.Steam]: DISCONNECTED_PROVIDER_ACCOUNT_STATE,
    [AppProvider.Gog]: DISCONNECTED_PROVIDER_ACCOUNT_STATE,
    [AppProvider.EpicGames]: DISCONNECTED_PROVIDER_ACCOUNT_STATE
  }
};
