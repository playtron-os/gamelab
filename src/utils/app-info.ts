import {
  AppDownloadStage,
  AppInformation,
  AppStatus,
  AppProvider
} from "@/types";

import { PlaytronImage, InstalledApp } from "@/types/app-library";
import { t } from "@lingui/macro";
import { CloudFill, RunFill } from "@playtron/styleguide";

export function getProviderName(provider: string): string {
  return (
    {
      [AppProvider.Steam]: "Steam",
      [AppProvider.Gog]: "GOG",
      [AppProvider.EpicGames]: "Epic Games"
    }[provider] || provider[0].toUpperCase() + provider.substring(1)
  );
}

export function nearestPowerOfTwo(n: number) {
  const log = Math.round(Math.log(n) / Math.log(2));
  const a = Math.round(Math.pow(2, log));
  const b = Math.round(Math.pow(2, log + 1));

  return n - a < b - n ? a : b;
}

export function getDiskSize(size: number | undefined): string {
  if (!size) {
    return "";
  }
  const units = ["B", "KB", "MB", "GB"];
  let unitIndex = 0;
  while (size > 1024 && unitIndex < units.length - 1) {
    size /= 1024;
    ++unitIndex;
  }
  return `${Math.round(size * 10) / 10}${units[unitIndex]}`;
}

/**
 * Variant of getDiskSize to return size printed
 * on the disk's label such as 2TB or 256GB.
 */
export function getDiskSizeLabel(size: number): string {
  if (size === 0) {
    return "";
  }
  const units = ["B", "KB", "MB", "GB", "TB"];
  let unitIndex = 0;
  while (size > 990 && unitIndex < units.length - 1) {
    size /= 1000;
    ++unitIndex;
  }
  const diskSize = Math.round(Math.round(size * 100) / 100);
  const nearestSize = nearestPowerOfTwo(diskSize);
  const sizeRatio = diskSize / nearestSize;
  if (sizeRatio >= 0.99 || sizeRatio <= 1.01) {
    return `${nearestSize}${units[unitIndex]}`;
  }
  return `${diskSize}${units[unitIndex]}`;
}

export const getAppStatusLabel = (status: AppStatus) => {
  switch (status) {
    case AppStatus.READY:
      return t`Ready`;
    case AppStatus.PRE_ALLOCATING:
      return t`Pre-Allocating`;
    case AppStatus.VERIFIYING:
      return t`Verifying`;
    case AppStatus.DOWNLOADING:
      return t`Downloading`;
    case AppStatus.INSTALLING:
      return t`Installing`;
    case AppStatus.UPDATE_REQUIRED:
      return t`Update Required`;
    case AppStatus.PAUSED:
      return t`Paused`;
    case AppStatus.QUEUED:
      return t`Queued`;
    case AppStatus.NOT_DOWNLOADED:
      return t`Not Downloaded`;
    case AppStatus.RUNNING:
      return t`Running`;
    case AppStatus.LAUNCHING:
      return t`Launching`;
    default:
      return t`Unknown`;
  }
};

export const getProgress = (installedApp?: InstalledApp): number => {
  if (!installedApp) return 0;
  switch (installedApp.download_status.stage) {
    case AppDownloadStage.DOWNLOADING:
    case AppDownloadStage.PRE_ALLOCATING:
    case AppDownloadStage.VERIFYING:
      return installedApp.download_status.progress || 0;
    case AppDownloadStage.INSTALLING:
      return (
        (installedApp.post_install.reduce(
          (prev, val) => prev + (val.is_success ? 1 : 0),
          0
        ) /
          installedApp.post_install.length) *
        100
      );
    default:
      return 0;
  }
};

export function getImage(images: PlaytronImage[]): string {
  if (!images) {
    return "";
  }
  let gogCandidate = "";
  let steamCandidate = "";
  let epicGamesCandidate = "";
  for (const image of images) {
    if (image.source === "steam" && image.image_type === "header") {
      steamCandidate = image.url;
    }

    if (
      image.source === "gog" &&
      image.url != "https" &&
      image.image_type === "logo"
    ) {
      gogCandidate = image.url.replace("https//", "https://");
    }
    if (image.source === "epicgames" && image.image_type === "OfferImageWide") {
      epicGamesCandidate = image.url;
    }
  }
  if (gogCandidate) {
    return gogCandidate;
  }
  if (epicGamesCandidate) {
    return epicGamesCandidate;
  }
  if (steamCandidate) {
    return steamCandidate;
  }

  return images.length.toString();
}

export function getDate(date: Date | number): string {
  return new Date(date).toDateString();
}

export const getAppActionLabelByStatus = (status: AppStatus) => {
  switch (status) {
    case AppStatus.READY:
      return t`Launch`;
    case AppStatus.PRE_ALLOCATING:
    case AppStatus.VERIFIYING:
    case AppStatus.DOWNLOADING:
      return t`Pause`;
    case AppStatus.NOT_DOWNLOADED:
      return t`Download`;
    case AppStatus.UPDATE_REQUIRED:
      return t`Update`;
    case AppStatus.PAUSED:
      return t`Resume`;
    case AppStatus.QUEUED:
      return t`Prioritize`;
    case AppStatus.RUNNING:
      return t`Stop`;
    case AppStatus.LAUNCHING:
      return t`Launching...`;
    default:
      return "...";
  }
};

/**
 * Get the App status without the QUEUED status which is not accessible from the AppInformation object
 * To get the AppStatus including the QUEUED status use getAppStatusWithQueue.
 *
 * @param appInfo
 * @returns AppStatus
 */
export const getAppStatus = (appInfo: AppInformation): AppStatus => {
  if (appInfo.is_running) {
    return AppStatus.RUNNING;
  } else if (appInfo.is_launched) {
    return AppStatus.LAUNCHING;
  }

  if (appInfo.is_installing) {
    return AppStatus.INSTALLING;
  }

  if (appInfo.is_downloading) {
    switch (appInfo.installed_app?.download_status.stage) {
      case AppDownloadStage.PRE_ALLOCATING:
        return AppStatus.PRE_ALLOCATING;
      case AppDownloadStage.VERIFYING:
        return AppStatus.VERIFIYING;
      default:
        return AppStatus.DOWNLOADING;
    }
  }

  switch (appInfo.installed_app?.download_status.stage) {
    case AppDownloadStage.PRE_ALLOCATING:
    case AppDownloadStage.VERIFYING:
    case AppDownloadStage.DOWNLOADING:
      return AppStatus.PAUSED;
    case AppDownloadStage.UPDATE_REQUIRED:
      return AppStatus.UPDATE_REQUIRED;
    case AppDownloadStage.DONE:
      return AppStatus.READY;
    default:
      return AppStatus.NOT_DOWNLOADED;
  }
};

/**
 * Return AppStatus, including QUEUED. Make sure to provide the necessary React context.
 * @param appInfo
 * @param queuePositionMapState
 * @returns AppStatus
 */
export const getAppStatusWithQueue = (
  appInfo: AppInformation,
  queuePositionMapState: { [key: string]: number }
): AppStatus => {
  const status = getAppStatus(appInfo);
  if (
    [
      AppStatus.RUNNING,
      AppStatus.INSTALLING,
      AppStatus.DOWNLOADING,
      AppStatus.PRE_ALLOCATING,
      AppStatus.VERIFIYING
    ].includes(status)
  ) {
    return status;
  }
  if (
    appInfo.installed_app &&
    typeof queuePositionMapState[appInfo.installed_app.owned_app.id] ===
      "number"
  ) {
    return AppStatus.QUEUED;
  }
  return status;
};

export const getAppActionIconByStatus = (status: AppStatus) => {
  if (status === AppStatus.READY) {
    return RunFill;
  }
  if (status === AppStatus.UPDATE_REQUIRED) {
    return CloudFill;
  }
};

/**
 * Return a user friendly label for a drive.
 * Uses the last part of the mount point or return "System Drive" for the root mount point.
 * @param drive
 * @returns string
 */
export const getDriveLabel = (drive: string | undefined): string => {
  if (!drive) {
    return "";
  }
  if (drive === "/") {
    return t`System Drive`;
  }

  const label = drive.split("/").pop() || drive;
  return label.replace(".ext4", "");
};
