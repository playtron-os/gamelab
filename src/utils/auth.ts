import { getFromLocalStorage, setInLocalStorage } from "./local-storage";

export const setUserIdToLocalStorage = (userId?: string) => {
  return setInLocalStorage("user_id", userId);
};

export const getUserIdFromLocalStorage = () => {
  return getFromLocalStorage("user_id");
};
