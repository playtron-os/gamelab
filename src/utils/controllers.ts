import { ControllerInfo, InputEvent } from "@/types/input-config";
import { t } from "@lingui/macro";
import { physicalLayouts } from "@/constants/physical-layouts";

export const KEYBOARD_QWERTY_LAYOUT = [
  [
    "KeyEsc",
    "KeyF1",
    "KeyF2",
    "KeyF3",
    "KeyF4",
    "KeyF5",
    "KeyF6",
    "KeyF7",
    "KeyF8",
    "KeyF9",
    "KeyF10",
    "KeyF11",
    "KeyF12"
  ],
  [
    "KeyGrave",
    "Key1",
    "Key2",
    "Key3",
    "Key4",
    "Key5",
    "Key6",
    "Key7",
    "Key8",
    "Key9",
    "Key0",
    "KeyMinus",
    "KeyEqual",
    "KeyBackspace"
  ],
  [
    "KeyTab",
    "KeyQ",
    "KeyW",
    "KeyE",
    "KeyR",
    "KeyT",
    "KeyY",
    "KeyU",
    "KeyI",
    "KeyO",
    "KeyP",
    "KeyLeftBrace",
    "KeyRightBrace",
    "KeyBackslash"
  ],
  [
    "KeyCapslock",
    "KeyA",
    "KeyS",
    "KeyD",
    "KeyF",
    "KeyG",
    "KeyH",
    "KeyJ",
    "KeyK",
    "KeyL",
    "KeySemicolon",
    "KeyApostrophe",
    "KeyEnter"
  ],
  [
    "KeyLeftShift",
    "KeyZ",
    "KeyX",
    "KeyC",
    "KeyV",
    "KeyB",
    "KeyN",
    "KeyM",
    "KeyComma",
    "KeyDot",
    "KeySlash",
    "KeyRightShift"
  ],
  [
    "KeyLeftCtrl",
    "KeyLeftMeta",
    "KeyLeftAlt",
    "KeySpace",
    "KeyRightAlt",
    "KeyCompose",
    "KeyRightMeta",
    "KeyRightCtrl"
  ]
];

export const KEYBOARD_NUMPAD_LAYOUT = [
  [
    "KeyInsert",
    "KeyHome",
    "KeyPageUp",
    "KeyPlay",
    "KeyStop",
    "KeyNext",
    "KeyPause"
  ],
  [
    "KeyDelete",
    "KeyEnd",
    "KeyPageDown",
    "KeyEqual",
    "KeySlash",
    "KeyAsterisk",
    "KeyMinus"
  ],
  [null, null, null, "KeyKp7", "KeyKp8", "KeyKp9", "KeyPlus"],
  [null, null, null, "KeyKp4", "KeyKp5", "KeyKp6"],
  [null, "KeyUp", null, "KeyKp1", "KeyKp2", "KeyKp3", "KeyEnter"],
  ["KeyLeft", "KeyDown", "KeyRight", "KeyKp0", "KeyDot"]
];

export const getKeyWidth = (key: string) => {
  switch (key) {
    case "KeyGrave":
      return "w-[68px]";
    case "KeyBackspace":
      return "w-[76px]";
    case "KeyTab":
      return "w-[96px]";
    case "KeyEsc":
    case "KeyF1":
    case "KeyF2":
    case "KeyF3":
    case "KeyF4":
    case "KeyF5":
    case "KeyF6":
    case "KeyF7":
    case "KeyF8":
    case "KeyF9":
    case "KeyF10":
    case "KeyF11":
    case "KeyF12":
      return "w-[55.71px]";
    case "KeyCapslock":
      return "w-[111px]";
    case "KeyRightShift":
    case "KeyLeftShift":
      return "w-[124px]";
    case "KeyEnter":
      return "w-[85px]";
    case "KeyRightCtrl":
    case "KeyLeftCtrl":
    case "KeyRightMeta":
    case "KeyLeftMeta":
    case "KeyCompose":
    case "KeyRightAlt":
    case "KeyLeftAlt":
      return "w-[65px]";
    case "KeySpace":
      return "w-[289px]";
    default:
      return "w-12";
  }
};

export const getKeyLabel = (key: string | null) => {
  if (!key) {
    return t`Unset`;
  }
  switch (key) {
    case "KeyGrave":
      return "`";
    case "KeySpace":
      return "Space";
    case "KeyBackspace":
      return "⇐";
    case "KeyEnter":
      return "⏎";
    case "KeyLeftShift":
    case "KeyRightShift":
      return "Shift";
    case "KeyLeftMeta":
    case "KeyRightMeta":
      return "Win";
    case "KeyLeftCtrl":
    case "KeyRightCtrl":
      return "Ctrl";
    case "KeyRightAlt":
    case "KeyLeftAlt":
      return "Alt";
    case "KeyLeftBrace":
      return "[";
    case "KeyRightBrace":
      return "]";
    case "KeyEqual":
      return "=";
    case "KeyComma":
      return ",";
    case "KeyDot":
      return ".";
    case "KeySlash":
      return "/";
    case "KeyBackslash":
      return "\\";
    case "KeyCompose":
      return "Fn";
    case "KeyApostrophe":
      return "'";
    case "KeySemicolon":
      return ";";
    case "KeyMinus":
      return "-";
    case "KeyInsert":
      return "Ins";
    case "KeyDelete":
      return "Del";
    case "KeyPageUp":
      return "PgUp";
    case "KeyPageDown":
      return "PgDn";
    case "KeyPlus":
      return "+";
    case "KeyAsterisk":
      return "*";
    case "KeyUp":
      return "↑";
    case "KeyDown":
      return "↓";
    case "KeyLeft":
      return "←";
    case "KeyRight":
      return "→";
    default:
      return key.replace("KeyKp", "").replace("Key", "");
  }
};

export const getInputLabel = (input: InputEvent, layout: string) => {
  const fixName = (name: string) => {
    return name.replace("Pad", " Pad");
  };

  if (!input) {
    return t`Unset`;
  }

  if (input.keyboard) {
    return getKeyLabel(input.keyboard);
  }
  if (input.mouse) {
    if (input.mouse.motion) {
      return t`Mouse Move`;
    }
    switch (input.mouse.button) {
      case "Left":
        return "Left Click";
      case "Right":
        return "Right Click";
      case "Middle":
        return "Middle Click";
      case "Extra1":
        return "Mouse Button 4";
      case "Extra 2":
        return "Mouse button 5";
      default:
        return input.mouse.button;
    }
  }
  if (input.touchpad) {
    if (input.touchpad.touch.button) {
      switch (input.touchpad.touch.button) {
        case "Press":
          return fixName(input.touchpad.name) + " Press";
        case "Touch":
          return fixName(input.touchpad.name) + " Touch";
      }
    }
    if (input.touchpad.touch.motion) {
      return fixName(input.touchpad.name) + " Move";
    }
  }
  if (input.gamepad) {
    if (input.gamepad.axis) {
      switch (input.gamepad.axis.name) {
        case "LeftStick":
          return "Left Stick";
        case "RightStick":
          return "Right Stick";
        default:
          return input.gamepad.axis.name;
      }
    }
    if (input.gamepad.trigger) {
      switch (input.gamepad.trigger.name) {
        case "LeftTrigger":
          return "Left Trigger";
        case "RightTrigger":
          return "Right Trigger";
        default:
          return input.gamepad.trigger.name;
      }
    }
    if (!input.gamepad.button) {
      return "?";
    }
    switch (input.gamepad.button) {
      case "LeftStick":
        return "L3 Button";
      case "RightStick":
        return "R3 Button";
      case "LeftBumper":
        return "Left Bumper";
      case "RightBumper":
        return "Right Bumper";
      default:
        return convertOrientationToButton(input.gamepad.button, layout);
    }
  }
  return "-";
};

/**
 * Given a button position, return the associated button label for a given controller.
 */
export const convertOrientationToButton = (label: string, layout?: string) => {
  if (layout === "ps5") {
    switch (label) {
      case "Start":
        return "Options";
      case "Select":
        return "Create";
      case "North":
        return "Square";
      case "West":
        return "Triangle";
      case "East":
        return "Circle";
      case "South":
        return "Cross";
      default:
        return label;
    }
  }
  if (layout === "xbox") {
    switch (label) {
      case "North":
        return "X";
      case "West":
        return "Y";
      case "East":
        return "B";
      case "South":
        return "A";
      default:
        return label;
    }
  }
  return label;
};

export const mappingCmp = (mapping1?: InputEvent, mapping2?: InputEvent) => {
  if (!mapping1 || !mapping2) {
    return false;
  }

  if (mapping1.keyboard && mapping2.keyboard) {
    return mapping1.keyboard === mapping2.keyboard;
  }
  if (mapping1.mouse && mapping2.mouse) {
    if ("button" in mapping1.mouse && "button" in mapping2.mouse) {
      return mapping1.mouse.button === mapping2.mouse.button;
    }
    if ("motion" in mapping1.mouse && "motion" in mapping2.mouse) {
      return true;
    }
  }
  if (mapping1.touchpad && mapping2.touchpad) {
    if (mapping1.touchpad.name !== mapping2.touchpad.name) {
      return false;
    }
    if (
      "button" in mapping1.touchpad.touch &&
      "button" in mapping2.touchpad.touch
    ) {
      return (
        mapping1.touchpad.touch.button === mapping2.touchpad.touch.button &&
        mapping1.touchpad.name === mapping2.touchpad.name
      );
    }
    if (
      "motion" in mapping1.touchpad.touch &&
      "motion" in mapping2.touchpad.touch
    ) {
      return true;
    }
  }
  if (mapping1.gamepad && mapping2.gamepad) {
    if ("button" in mapping1.gamepad && "button" in mapping2.gamepad) {
      return mapping1.gamepad.button === mapping2.gamepad.button;
    }
    if (mapping1.gamepad.trigger && mapping2.gamepad.trigger) {
      return mapping1.gamepad.trigger.name === mapping2.gamepad.trigger.name;
    }
    if (mapping1.gamepad.axis && mapping2.gamepad.axis) {
      if (mapping1.gamepad.axis.direction !== mapping2.gamepad.axis.direction) {
        return false;
      }
      return mapping1.gamepad.axis.name === mapping2.gamepad.axis.name;
    }
  }
  return false;
};

export const getPhysicalLayoutFromDevice = (device: ControllerInfo) => {
  // Xbox360
  if (device.capabilities.length === 20) {
    return physicalLayouts.Xbox;
  }
  if (device.capabilities.length === 21) {
    return physicalLayouts.Xbox;
  }
  // Xbox One
  if (device.capabilities.length === 25) {
    return physicalLayouts.Xbox;
  }
  if (device.capabilities.length === 33) {
    return physicalLayouts.PS5;
  }
  // Aya Neo 2
  // if (device.capabilities.length === 155) {
  //   return physicalLayouts.SteamDeck;
  // }
  if (device.capabilities.length === 173) {
    return physicalLayouts.SteamDeck;
  }
  if (device.capabilities.length === 186) {
    return physicalLayouts.ROGAlly;
  }
  console.log("Unknown device", device);
  console.log("Device capabilities", device.capabilities);
  return physicalLayouts.Generic;
};
