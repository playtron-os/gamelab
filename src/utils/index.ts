export * from "./auth";
export * from "./test-utils";
