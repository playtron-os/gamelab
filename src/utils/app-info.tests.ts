import { nearestPowerOfTwo } from "./app-info";

describe("Disk size", () => {
  test("Nearest power of two", () => {
    expect(nearestPowerOfTwo(510)).toBe(512);
    expect(nearestPowerOfTwo(513)).toBe(512);
    expect(nearestPowerOfTwo(63)).toBe(64);
    expect(nearestPowerOfTwo(64)).toBe(64);
  });
});
