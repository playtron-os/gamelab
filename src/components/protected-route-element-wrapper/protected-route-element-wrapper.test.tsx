import React from "react";
import { ProtectedRouteElementWrapper } from ".";
import { screen } from "@testing-library/react";
import { renderWithAllContext } from "@/utils";
import { configureStore } from "@reduxjs/toolkit";
import {
  APP_LIBRARY_INITIAL_STATE,
  accountsSlice,
  appLibrarySlice,
  authSlice
} from "@/redux/modules";
import { PROVIDERS_ACCOUNTS_DISCONNECTED_STATE_MOCK } from "@/mocks/provider-account-state";

const getMockStore = ({
  isAuthenticated = false
}: {
  isAuthenticated?: boolean;
}) => {
  return configureStore({
    reducer: {
      appLibrary: appLibrarySlice.reducer,
      auth: authSlice.reducer,
      accounts: accountsSlice.reducer
    },
    preloadedState: {
      appLibrary: APP_LIBRARY_INITIAL_STATE,
      auth: {
        isAuthenticated,
        userId: ""
      },
      accounts: PROVIDERS_ACCOUNTS_DISCONNECTED_STATE_MOCK
    }
  });
};

const TestComponent = () => (
  <ProtectedRouteElementWrapper>Test component</ProtectedRouteElementWrapper>
);

describe("ProtectedRouteElementWrapper", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("snapshot", () => {
    const { asFragment } = renderWithAllContext(
      <TestComponent />,
      undefined,
      getMockStore({})
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders children when user is authenticated", () => {
    renderWithAllContext(
      <TestComponent />,
      undefined,
      getMockStore({
        isAuthenticated: true
      })
    );

    expect(screen.getByText("Test component")).toBeInTheDocument();
  });

  it("redirects to /auth when user is not authenticated", () => {
    renderWithAllContext(<TestComponent />, undefined, getMockStore({}));
  });
});
