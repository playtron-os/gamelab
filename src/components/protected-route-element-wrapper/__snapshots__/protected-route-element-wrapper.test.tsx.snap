// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRouteElementWrapper snapshot 1`] = `<DocumentFragment />`;
